Parser/rdiBeam2Earth.m

```diff
@@ -0,0 +1,194 @@
+% ba is the beam angle, in degrees.
+% cvxccv is 1 for a convex xdcr, other than 1 for a concave xdcr.  Usually 1.
+% updwon is 0 for a down-looking instrument, other than 0 for an up-looker.
+%
+% Calls matlab routines beam2inst.m and inst2earth.m.
+%
+head = head*0.01;
+pitch = pitch *0.01;
+roll = roll * 0.01;
+% make pitch, roll, heading the same size as b1, b2, b3, b4
+elev=[-70 -70 -70 -70]; % default elev for 20o
+azi=[270 90 0 180];
+if updown == 1 %uplooking
+    txtup = 'up';
+else
+    txtup = 'down';
+end
+[veast,vnrth,wvel,evel] = deal(NaN*b1);
+% have to bin map it first. Even if the EX bit is set for bin
+% mapping, if the data is collected in beam coordinates, bin mapping
+% does not occur on board.
+[b1,b2,b3,b4] = binmap(b1,b2,b3,b4,head,pitch,roll,ba,distance);
+for a = 1:length(pitch)
+    beamdat = [b1(a,:); b2(a,:); b3(a,:); b4(a,:)];
+    thePitch=atand(tand(pitch(a))*cosd(roll(a))); % CBluteau as per RDI coordinate transformation manual Eq 19 when EZxxx1xxx is set
+    i2e=inst2earth(head(a),thePitch,roll(a),updown);
+    b2i=beam2inst(ba,cvxccv);
+    V = (i2e*b2i*beamdat)'; %V = [u v w V_err];
+    veast(a,:) = V(:,1)';
+    vnrth(a,:) = V(:,2)';
+    wvel(a,:) = V(:,3)';
+    evel(a,:) = V(:,4)';
+end
+return
+end
+
+function y=inst2earth(head,pitch,roll,updown)
+% Creates transformation matrix for converting from instrument coordinates to
+% earth coordinates based on heading, pitch and roll information.  Assumes
+% arguments are given in degrees.  Updown should be 0 for a down-looking system,
+% and some number other than 0 for an up-looking system.
+%
+% Note that the matrix created is 4x4, with the additional row and column added
+% so that the error velocity calculated in beam2inst.m will carry through.
+%
+% Convert input degrees to radians
+hrad=head*pi/180;
+prad=pitch*pi/180;
+if updown == 0
+    rrad=roll*pi/180;
+else
+    rrad=(roll+180)*pi/180;
+end
+zero= 0;
+one = 1;
+%
+% Create the heading rotation matrix:
+%
+h=[cos(hrad)  sin(hrad)  zero zero
+    -sin(hrad) cos(hrad) zero zero
+    zero zero one zero
+    zero zero zero one];
+%
+% Create the pitch rotation matrix:
+%
+p=[one zero zero zero
+    zero  cos(prad) -sin(prad) zero
+    zero  sin(prad) cos(prad) zero
+    zero zero zero one];
+%
+% Create the roll rotation matrix:
+%
+r=[cos(rrad) zero sin(rrad) zero
+    zero  one  zero zero
+    -sin(rrad) zero cos(rrad) zero
+    zero zero zero one];
+%
+% create the final matrix
+%
+y=h*p*r;
+return
+end
+
+function y=beam2inst(ba,c)
+% Creates transformation matrix for converting from beam coordinates to
+% instrument coordinates based the beam angle.
+%
+% ba is the beam angle of the ADCP (typically either 20 or 30 degrees)
+% c is whether the instrument is convex (c=1) or concave (c other than 1). Nearly
+% all ADCPs are convex (some early vessel-mounted systems were concavbe).
+%
+% Create the apropriate scale factors:
+%
+barad=ba*pi/180;
+a=1/(2*sin(barad));
```

Comment:
same as above, lets use the same units in the functionality and avoid unecessary operations.

Reply:
I also wonder if we could use symbolic maths here for more concise execution.